qbft/round_change.go

```diff
@@ -27,40 +22,44 @@ func (i *Instance) uponRoundChange(
 		i.State,
 		i.config,
 		signedRoundChange,
-		roundChangeMsgContainer,
-		valCheck)
+		i.State.RoundChangeContainer,
+		i.config.GetValueCheckF())
 	if err != nil {
-		return errors.Wrap(err, "could not get proposal justification for leading ronud")
+		return errors.Wrap(err, "could not get proposal justification for leading round")
 	}
 	if justifiedRoundChangeMsg != nil {
-		highestRCData, err := justifiedRoundChangeMsg.Message.GetRoundChangeData()
-		if err != nil {
-			return errors.Wrap(err, "could not round change data from highestJustifiedRoundChangeMsg")
-		}
-
 		// Chose proposal value.
 		// If justifiedRoundChangeMsg has no prepare justification chose state value
 		// If justifiedRoundChangeMsg has prepare justification chose prepared value
-		valueToPropose := instanceStartValue
-		if highestRCData.Prepared() {
-			valueToPropose = highestRCData.PreparedValue
+		valueToPropose := i.StartValue
+		if justifiedRoundChangeMsg.Message.Prepared() {
+			// TODO<olegshmuelov>: validate that justified round change msg holds the complete input data
+			valueToPropose = &Data{
+				Root:   justifiedRoundChangeMsg.Message.InputRoot,
+				Source: justifiedRoundChangeMsg.InputSource,
+			}
 		}
```

Comment:
[Validate justified prepared round change msg holds the complete input data](https://github.com/bloxapp/ssv-spec/issues/103)